src/udev/hidraw_id/hidraw_id.c

```diff
@@ -0,0 +1,114 @@
+/* SPDX-License-Identifier: GPL-2.0+ */
+/*
+ * Modified from "u2f_hidraw_id.c" on 2019-08-19
+ *
+ * Original file header:
+ *
+ * Copyright (c) 2014-2015 Andrew Lutomirski
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Library General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Library General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <fcntl.h>
+#include <libudev.h>
+#include <linux/hid.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <syslog.h>
+#include <unistd.h>
+
+#include "fd-util.h"
+#include "hidraw_id_desc.h"
+#include "libudev-util.h"
+#include "log.h"
+#include "macro.h"
+#include "string-util.h"
+#include "udev-util.h"
+
+int main(int argc, char **argv) {
+        _cleanup_(udev_unrefp) struct udev *udev = NULL;
+        _cleanup_(udev_device_unrefp) struct udev_device *dev = NULL;
+        _cleanup_free_ char *path = NULL;
+        _cleanup_close_ int fd = -1;
+
+        struct udev_device *hiddev;
+        uint8_t desc[HID_MAX_DESCRIPTOR_SIZE];
+        int desclen;
+        struct HidrawIdDescResult desc_result = {0};
+
+        int ret = EXIT_FAILURE;
+
+        log_set_target(LOG_TARGET_AUTO);
+        udev_parse_config();
+        log_parse_environment();
+        log_open();
+
+        if (argc != 2) {
+                fprintf(stderr, "Usage: hidraw_id SYSFS_PATH|--udev\n");
+                goto out;
+        }
+
+        udev = udev_new();
+        if (!udev) {
+                log_oom();
+                goto out;
+        }
+
+        if (streq(argv[1], "--udev"))
+                dev = udev_device_new_from_environment(udev);
+        else
+                dev = udev_device_new_from_syspath(udev, argv[1]);
+        if (!dev)
+                goto out;
+
+        hiddev = udev_device_get_parent(dev);
+        if (!hiddev)
+                goto out;
+
+        path = strjoin(udev_device_get_syspath(hiddev), "/report_descriptor");
+        if (!path) {
+                log_oom();
+                goto out;
+        }
+
+        fd = open(path, O_RDONLY | O_NOFOLLOW | O_CLOEXEC);
+        if (fd == -1) {
+                log_debug("Failed to open report descriptor at: %s", path);
+                goto out;
+        }
+
+        desclen = read(fd, desc, sizeof(desc));
+        if (desclen <= 0)
+                goto out;
+
+        if (parse_hid_report_desc(desc, desclen, &desc_result) != 0) {
```

Comment:
We usually returns negative errno on error, and please do not ignore the returned value.
```
r = parse_hid_report_desc();
if (r < 0) {
    log_debug_errno(r, "Failed to parse report descriptor at '%s': %m", path);
    goto out;
}
```